notification/controller.go

```diff
@@ -0,0 +1,182 @@
+}
+
+type subscriber struct {
+	client      chan Event
+	eventTypes  []EventType
+	full        bool
+	lastEventID string
+}
+
+func NewController(s Storage) *Controller {
+	c := &Controller{
+		s:                    s,
+		Notifier:             make(chan Event, 1),
+		subscribingClients:   make(chan subscriber),
+		unsubscribingClients: make(chan chan Event),
+		activeClients:        make(map[chan Event]subscriber),
+		shutdown:             make(chan bool),
+	}
+	go c.handler()
+	return c
+}
+
+func (c *Controller) subscribe(client chan Event, eventTypes []EventType, full bool, lastEventID string) error {
+	s := subscriber{client: client,
+		eventTypes:  eventTypes,
+		full:        full,
+		lastEventID: lastEventID,
+	}
+	c.subscribingClients <- s
+	return nil
+}
+
+func (c *Controller) unsubscribe(client chan Event) error {
+	c.unsubscribingClients <- client
+	return nil
+}
+
+func (c *Controller) storeAndNotify(event Event) error {
+	var err error
+	event.ID, err = c.s.getNewID()
+	if err != nil {
+		return fmt.Errorf("error generating ID : %v", err)
+	}
+
+	// Notify
+	c.Notifier <- event
+
+	// Store
+	err = c.s.add(event)
+	if err != nil {
+		return fmt.Errorf("error storing the notification : %v", err)
+	}
+
+	return nil
+}
+
+func (c *Controller) Stop() {
+	c.shutdown <- true
+}
+
+func (c *Controller) CreateHandler(new catalog.ThingDescription) error {
+	event := Event{
+		Type: createEvent,
+		Data: new,
+	}
+
+	err := c.storeAndNotify(event)
+	return err
+}
+
+func (c *Controller) UpdateHandler(old catalog.ThingDescription, new catalog.ThingDescription) error {
+	oldJson, err := json.Marshal(old)
+	if err != nil {
+		return fmt.Errorf("error marshalling old TD")
+	}
+	newJson, err := json.Marshal(new)
+	if err != nil {
+		return fmt.Errorf("error marshalling new TD")
+	}
+	patch, err := jsonpatch.CreateMergePatch(oldJson, newJson)
+	if err != nil {
+		return fmt.Errorf("error merging new TD")
+	}
+	var td catalog.ThingDescription
+	if err := json.Unmarshal(patch, &td); err != nil {
+		return fmt.Errorf("error unmarshalling the patch TD")
+	}
+	td[wot.KeyThingID] = old[wot.KeyThingID]
+	event := Event{
+		Type: updateEvent,
+		Data: td,
+	}
+	err = c.storeAndNotify(event)
+	return err
+}
+
+func (c *Controller) DeleteHandler(old catalog.ThingDescription) error {
+	deleted := catalog.ThingDescription{
+		wot.KeyThingID: old[wot.KeyThingID],
+	}
+	event := Event{
+		Type: deleteEvent,
+		Data: deleted,
+	}
+	err := c.storeAndNotify(event)
+	return err
+}
+
+func (c *Controller) handler() {
+loop:
+	for {
+		select {
+		case s := <-c.subscribingClients:
+			c.activeClients[s.client] = s
+			log.Printf("New subscription. %d active clients", len(c.activeClients))
+
+			// Send the missed events
+			if s.lastEventID != "" {
+				missedEvents, err := c.s.getAllAfter(s.lastEventID)
+				if err != nil {
+					log.Printf("error getting the events after ID %s: %s", s.lastEventID, err)
+				}
+				for _, event := range missedEvents {
```

Comment:
`missedEvents` will be empty if err is not nil? Better put this in an `else` statement.